skl/ag-parent/src/main/java/se/inera/intyg/common/agparent/model/converter/certificate/AbstractQuestionDiagnoser.java

```diff
@@ -0,0 +1,198 @@
+                CertificateDataValueDiagnosisList.builder()
+                    .list(createDiagnosValue(diagnoser))
+                    .build()
+            )
+            .validation(
+                new CertificateDataValidation[]{
+                    CertificateDataValidationMandatory.builder()
+                        .questionId(questionId)
+                        .expression(withCitation(singleExpression(DIAGNOSES_LIST_ITEM_1_ID)))
+                        .build(),
+                    CertificateDataValidationText.builder()
+                        .limit(LIMIT_DIAGNOSIS_DESC)
+                        .build(),
+                    CertificateDataValidationShow.builder()
+                        .questionId(showValidationQuestionId)
+                        .expression(singleExpression(showValidationJsonId))
+                        .build()
+                }
+            )
+            .build();
+    }
+
+    private static List<CertificateDataValueDiagnosis> createDiagnosValue(List<Diagnos> diagnoses) {
+        if (diagnoses == null) {
+            return Collections.emptyList();
+        }
+
+        final List<CertificateDataValueDiagnosis> newDiagnoses = new ArrayList<>();
+        for (int i = 0; i < diagnoses.size(); i++) {
+            final var diagnosis = diagnoses.get(i);
+            if (diagnosis.getDiagnosKod() == null) {
+                continue;
+            }
+
+            newDiagnoses.add(createDiagnosis(Integer.toString(i + 1), diagnosis));
+        }
+
+        return newDiagnoses;
+    }
+
+    private static CertificateDataValueDiagnosis createDiagnosis(String id, Diagnos diagnos) {
+        return CertificateDataValueDiagnosis.builder()
+            .id(convertIntToId(id))
+            .terminology(diagnos.getDiagnosKodSystem())
+            .code(diagnos.getDiagnosKod())
+            .description(diagnos.getDiagnosBeskrivning())
+            .build();
+    }
+
+    protected static List<Diagnos> toInternal(Certificate certificate, String questionId, WebcertModuleService moduleService) {
+        var diagnosisList = diagnosisListValue(certificate.getData(), questionId);
+        int maxDiagnosIndex = getMaxDiagnosIndex(diagnosisList);
+
+        List<Diagnos> newDiagnosisList = new ArrayList<>();
+        while (newDiagnosisList.size() < maxDiagnosIndex) {
+            newDiagnosisList.add(Diagnos.create(null, null, null, null));
+        }
+
+        diagnosisList.forEach(diagnosis -> {
+            var newDiagnosis = Diagnos.create(
+                diagnosis.getCode(),
+                diagnosis.getTerminology(),
+                diagnosis.getDescription(),
+                moduleService.getDescriptionFromDiagnosKod(diagnosis.getCode(), diagnosis.getTerminology()));
+            newDiagnosisList.set(convertIdToInt(diagnosis.getId()), newDiagnosis);
+        });
+        return newDiagnosisList;
+    }
+
+    private static int getMaxDiagnosIndex(List<CertificateDataValueDiagnosis> diagnosisList) {
```

Comment:
Should it be called getMaxDiagnosisIndex instead?

Reply:
Good suggestion. Done.